### PR TITLE
state: Make parquet state management more reliable.

### PR DESCRIPTION
**Pull request by jacksonrnewhouse**
This change enables actually using the S3 backend for parquet and overhauls the logic and checkpoint management, file expiration and compaction. The main changes are

- Allow the Scheduler to pass environment variables down to the running workers. This lets it access the S3 backend.
- The ParquetFlusher is authoritative about which files are needed in a checkpoint. This will never change, and for GlobalTables it is always only the current epoch. All current usages align with this, rewriting the map at every checkpoint barrier, but you could imagine it changing in the future.
- All file deletion is managed by the controller as part of compaction. For epochs that are getting compacted it inspects the files in each epoch compared to the new min_epoch. Files not in min_epoch can be deleted.
- Deletion is always done in the following order:
1. Expired data files.
2. Operator metadata.
3. Checkpoint metadata.
- The BackingStore trait was updated to tolerate missing operator metadata. This makes it robust to a failure mid-compaction, which we didn't have before.

### CODE REVIEW

File: arroyo-rpc/proto/rpc.proto [hunk 1]
```diff
@@ -317,6 +317,18 @@ service WorkerGrpc {
...
+
+message StartPipelineReq {
+  string name = 1;
+  string pipeline_path = 2;
+  string wasm_path = 3;
+  string job_id = 4;
+  string hash = 5;
+  int64 run_id = 6;
+  uint64 slots = 7;
+  map<string, string> environment_variables = 8;
```
Comment (mwylde) [inline arroyo-rpc/proto/rpc.proto:330]: nit: we tend to shorten this to env_vars everywhere else in our code

File: arroyo-state/src/parquet.rs [hunk 2]
```diff
@@ -30,11 +30,11 @@ use tokio::fs::{remove_file, File};
 use tokio::io::{AsyncReadExt, AsyncWriteExt};
 use tokio::sync::mpsc::{self, Receiver, Sender};
 use tokio::sync::oneshot;
-use tracing::log::warn;
+use tracing::warn;
 use tracing::{debug, info};
 
-pub const S3_STORAGE_ENGINE_ENV: &str = "S3_BUCKET";
-pub const S3_REGION_ENV: &str = "S3_REGION";
+pub const S3_STORAGE_ENGINE_ENV: &str = "S3_STORAGE_BUCKET";
```
Comment (mwylde) [inline arroyo-state/src/parquet.rs:36]: Why is this a different env var than the user setting?

File: arroyo-controller/src/job_controller/mod.rs [hunk 3]
```diff
@@ -455,15 +455,14 @@ impl JobController {
...
-        if let Some(new_epoch) = self.model.compaction_needed() {
-            if self.compacting_task.is_none() && self.model.checkpoint_state.is_none() {
-                self.compacting_task = Some(self.start_compaction(new_epoch));
-            }
-        }
-
         // check on checkpointing
         if self.model.checkpoint_state.is_some() {
             self.model.finish_checkpoint_if_done(&self.pool).await?;
+            if let Some(new_epoch) = self.model.compaction_needed() {
```
[the remark this answers is not in the document]
Comment (jacksonrnewhouse) [reply arroyo-controller/src/job_controller/mod.rs:461]: Yes, because we don't wait for a "state loaded" type signal there was a race condition where the files a subtask needed could get compacted away. By deferring compaction we know we'll be far enough ahead on the running pipeline.

[the remark this answers is not in the document]
Comment (mwylde) [reply arroyo-controller/src/job_controller/mod.rs:461]: Ok, that seems like a bit of a hack around the actual problem but it's ok for now

File: arroyo-rpc/proto/rpc.proto [hunk 4]
```diff
@@ -317,6 +317,18 @@ service WorkerGrpc {
   rpc JobFinished(JobFinishedReq) returns (JobFinishedResp);
 }
 
+// Scheduler
+
+message StartPipelineReq {
```
[the remark this answers is not in the document]
Comment (jacksonrnewhouse) [reply arroyo-rpc/proto/rpc.proto:322]: Moving it out.

File: arroyo-rpc/proto/rpc.proto (hunk 1 above)
Comment (jacksonrnewhouse) [reply arroyo-rpc/proto/rpc.proto:330]: Will do.

File: arroyo-state/src/parquet.rs (hunk 2 above)
Comment (jacksonrnewhouse) [reply arroyo-state/src/parquet.rs:36]: Could imagine them being different, but probably not worth it right now. I'll put it back to S3_REGION

File: arroyo-controller/src/scheduler.rs [hunk 5]
```diff
@@ -643,6 +641,14 @@ impl Scheduler for NomadScheduler {
                     ]
                 }
             });
+            if let Some(env) = job["Job"]["TaskGroups"][0]["Tasks"][0]["Env"].as_object_mut() {
```
[the remark this answers is not in the document]
Comment (jacksonrnewhouse) [reply arroyo-controller/src/scheduler.rs:644]: Did this, converting everything to Strings is kind of a pain, let me know if you like it better.

File: arroyo-controller/src/scheduler.rs [hunk 6]
```diff
@@ -44,13 +45,7 @@ lazy_static! {
...
     async fn start_workers(
         &self,
-        pipeline_path: &str,
-        wasm_path: &str,
-        job_id: String,
-        run_id: i64,
-        name: String,
-        hash: String,
-        slots: usize,
+        start_pipeline_req: StartPipelineReq,
```
[the remark this answers is not in the document]
Comment (jacksonrnewhouse) [reply arroyo-controller/src/scheduler.rs:48]: It's not that big, just a few strings, and there are currently some Schedulers that move values into tokio::spawn calls, namely the ProcessScheduler.

File: arroyo-controller/src/job_controller/mod.rs [hunk 7]
```diff
@@ -24,7 +24,7 @@ use self::checkpointer::CheckpointState;
 
 mod checkpointer;
 
-const CHECKPOINTS_TO_KEEP: u32 = 2;
+const CHECKPOINTS_TO_KEEP: u32 = 4;
```
[the remark this answers is not in the document]
Comment (jacksonrnewhouse) [reply arroyo-controller/src/job_controller/mod.rs:27]: All the work is per checkpoint, as for each epoch we're compacting we check if they were the last to care about a file and, if so, drop it.

File: arroyo-controller/src/job_controller/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacksonrnewhouse) [reply arroyo-controller/src/job_controller/mod.rs:461]: Actually, looking again I think that this is no longer a problem, since we now rely on the invariant that GlobalTables are fully written to every checkpoint, which is currently true in all cases.

**Review by mwylde — APPROVED** [review]

**Review by mwylde — APPROVED** [review]
